### PR TITLE
Support signed custom headers in presigned URL generation

### PR DESCRIPTION
**Pull request by supernullset**
As discussed in #524 I discovered the hard way that I was unable to set s3 custom metadata. I opted for a Python [boto3](https://github.com/boto/botocore/blob/939fbebcae0e0519f040063d61f7dc354cd09dd8/botocore/signers.py#L240-L269) (via `prepare_request_dict/2` in `awsrequest.py`) style of metadata setter, for example/refresher:

python:
```
realurl = s3.generate_presigned_url(
    ClientMethod='put_object',
    Params={
        'Bucket': 'pharchive-dev',
        'Key': 'test.txt',
        'Metadata': {'metadata': 'foobar'}
    },
    HttpMethod='PUT',
)
```

Elixir:
```
custom_metadata = [{"metadata", "foobar"}]
ExAws.Auth.presigned_url(http_method, url, service, datetime, @config, expires, query_params, custom_metadata)
```

I generated a test using a very basic python boto3 script in order to generate the correct signatures to pull into the Elixir tests:

```
import boto3
from boto3 import Session
from botocore.client import Config

 session = Session(
     aws_access_key_id= 'AKIAIOSFODNN7EXAMPLE',
     aws_secret_access_key= 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY',
 )

# Get the service client
s3 = session.client('s3', config=Config(
    signature_version='s3v4'
))

realurl = s3.generate_presigned_url(
    ClientMethod='put_object',
    Params={
        'Bucket': 'pharchive-dev',
        'Key': 'test.txt',
        'Metadata': {'metadata': 'foobar'}
    },
    HttpMethod='PUT',
)

print(realurl)
```

This is an airity change, so Im not sure where else I would need to make updates in other libraries. I am happy to take on the job if you point me in the right direction @benwilson512. Thanks for the great library, and sorry for the delay in getting this PR together.

### CODE REVIEW

File: lib/ex_aws/auth.ex [hunk 1]
```diff
@@ -49,13 +49,13 @@ defmodule ExAws.Auth do
     end
   end
 
-  def presigned_url(http_method, url, service, datetime, config, expires, query_params \\ [], body \\ nil) do
+  def presigned_url(http_method, url, service, datetime, config, expires, query_params \\ [], custom_metadata \\ [], body \\ nil) do
```
Comment (benwilson512) [inline lib/ex_aws/auth.ex]: Hey! Happy to merge this, but we can't make a breaking change like this. Please re-locate the `custom_metadata` arg to be after `body`.

Comment (supernullset) [reply lib/ex_aws/auth.ex]: Sure! Happy to do it - let me know if anything else needs to be shuffled

**Comment by benwilson512** [comment]
Hey @supernullset if you can update the merge conflicts and make that change we can get this into the next version.

**Comment by supernullset** [comment]
@benwilson512 Changes updated - I am a little iffy about my changes to `presigned_url_headers` - I think there is a more elegant solution that is just escaping me at the moment. Sorry for the delay - I'll make sure to be available in the next week if there are more changes you would like to have made

**Comment by Doerge** [comment]
I'll politely ask for a status on this @benwilson512 .
Is there anything else that needs to be done?

**Comment by benwilson512** [comment]
Honestly the largest holdup is that there are competing PRs and issues with respect to URL signing. https://github.com/ex-aws/ex_aws/pull/603 and the whole discussion here https://github.com/ex-aws/ex_aws/issues/602 

I haven't had time to unravel what the correct behavior is supposed to be.

**Comment by brianlow** [comment]
Is this PR still needed? It looks like a variation of this is in master:

https://github.com/ex-aws/ex_aws/blob/master/lib/ex_aws/auth.ex

**Comment by bernardd** [comment]
> Is this PR still needed? It looks like a variation of this is in master:

Given that nobody has replied to this, and in the interests of cleaning up the PR list a bit, I'm going to close this. @supernullset please don't hesitate to reopen if this is still an issue.

**Pull request closed without merging**